packages/devkit/src/utils/package-json.ts

```diff
@@ -368,89 +372,109 @@ function requiresRemovingOfPackages(
-    const packageJson = readJson(tree, 'package.json');
-    const field = dev ? 'devDependencies' : 'dependencies';
-    version = packageJson[field]?.[pkg];
-  }
-
-  if (
-    // Special case: When running Nx unit tests, the version read from package.json is "0.0.1".
-    !(
-      pkg.startsWith('@nrwl/') &&
-      (version === '0.0.1' || requiredVersion === '0.0.1')
-    ) &&
-    // Normal case
-    !satisfies(version, requiredVersion, { includePrerelease: true })
-  ) {
-    const installCmd = `${
-      dev ? pmc.addDev : pmc.add
-    } ${pkg}@${requiredVersion}`;
-    if (throwOnMissing) {
-      throw new Error(
-        `Cannot install required package ${pkg} during a dry run. Run the generator without --dryRun, or install the package with "${installCmd}" and try again.`
-      );
-    } else {
-      execSync(installCmd, {
-        cwd: tree.root,
-        stdio: [0, 1, 2],
-      });
-    }
-  }
-}
+  options?: { dev?: boolean; throwOnMissing?: boolean }
+): void;
 
 /**
- * Use another process to resolve the package.json path of the package (if it exists).
- * Cannot use `require.resolve` here since there is an unclearable internal cache used by Node that can lead to issues
- * when resolving the package after installation.
+ * Ensure that dependencies and devDependencies from package.json are installed at the required versions.
  *
- * See: https://github.com/nodejs/node/issues/31803
+ * For example:
+ * ```typescript
+ * ensurePackage(tree, '@nrwl/jest', nxVersion)
+ * ```
+ * @param pkg the package to install and require
+ * @param version the version to install if the package doesn't exist already
  */
-function getPackageVersion(pkg: string): undefined | string {
+export function ensurePackage<T extends any = any>(
+  pkg: string,
+  version: string
+): T;
+export function ensurePackage<T extends any = any>(
+  pkgOrTree: string | Tree,
+  requiredVersionOrPackage: string,
+  maybeRequiredVersion?: string,
+  _?: never
+): T {
+  let pkg: string;
+  let requiredVersion: string;
+  if (typeof pkgOrTree === 'string') {
+    pkg = pkgOrTree;
+    requiredVersion = requiredVersionOrPackage;
+  } else {
+    // Old Signature
+    pkg = requiredVersionOrPackage;
+    requiredVersion = maybeRequiredVersion;
+  }
+
+  if (packageMapCache.has(pkg)) {
+    return packageMapCache.get(pkg) as T;
+  }
+
   try {
-    return execSync(
-      `node -e "console.log(require('${pkg}/package.json').version)"`,
-      {
-        cwd: workspaceRoot,
-        stdio: ['pipe', 'pipe', 'ignore'],
-      }
-    )
-      .toString()
-      .trim();
+    return require(pkg);
   } catch (e) {
-    return undefined;
+    if (e.code !== 'MODULE_NOT_FOUND') {
+      throw e;
+    }
   }
+
+  if (process.env.NX_DRY_RUN && process.env.NX_DRY_RUN !== 'false') {
+    throw new Error(
+      'NOTE: This generator does not support --dry-run. If you are running this in Nx Console, it should execute fine once you hit the "Run" button.\n'
+    );
+  }
```

Comment:
Do we still need this? It  seems like this part should actually work in dry run with the new mode since its not installing in the repo.